### PR TITLE
Switch dependency manager to go modules

### LINKED ISSUE #529: Switch to gomodules for dependency management
With introduction of gomodules in from GO 1.11, we should switch to go modules for dependency management.

### PR DESCRIPTION
**Pull request by kayrus**
**What this PR does / why we need it**:

Switch the dependency manager to go modules

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Resolves #529

**Special notes for your reviewer**:

**Release note**:

```release-note
Switch dependency manager to go modules
```

### CODE REVIEW

**Comment by lingxiankong** [comment]
@kayrus Hi, I found you usually submit multiple commits in a single PR which is not recommended for upstream contribution, at least in openstack-provider. 

If it's a big task(e.g. affects all the binaries) you are working on, I'd suggest splitting the task into multiple sub-tasks and submit multiple PRs while associating those PRs using the same issue number. Otherwise, please just submit the PR containing only one single commit.

I know it's painful compared to Gerrit, but this is what we have.

**Comment by kayrus** [comment]
@lingxiankong thanks for the note. From my side am still quite confused of your requirement, because github supports [squash and merge](https://help.github.com/en/articles/about-pull-request-merges). If you have concerns about squashed commit message(s) - the PR topic has to represent the squashed commit.

As for the current PR state. It looks like there should be some modifications on the openlab side. Probably the golang version has to be bumped to 1.13 as well. I tested the compilation process in the clean docker environment and using go modules I was able to compile the package without installing 3rd-party dependency management software.

**Comment by lingxiankong** [comment]
> because github supports [squash and merge](https://help.github.com/en/articles/about-pull-request-merges). If you have concerns about squashed commit message(s) - the PR topic has to represent the squashed commit.

Maybe I missed something, I know I can do `squash and merge` commits for the PRs in my own repo, but I am not sure what's the default behavior in openstack-provider when a PR with multiple commits is approved and how it's going to be merged. @adisky do you know if we could do `squash and merge` when approving PRs?

**Comment by adisky** [comment]
@linxiankong we support auto squash in this repo too
@kayrus thanks for your PR, ya need to update version in openlab also, let's wait for next cycle to get this PR in
/hold

**Comment by lingxiankong** [comment]
> @linxiankong we support auto squash in this repo too

Sounds good, that solves my concerns. Thanks!

File: .zuul.yaml [hunk 1]
```diff
@@ -78,7 +78,7 @@
         - cloud-provider-openstack-acceptance-test-e2e-conformance:
             files:
               - ^pkg/cloudprovider/.*
-              - ^Gopkg.toml$
```
Comment (ramineni) [inline .zuul.yaml]: we could add go.sum also here?

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,87 @@
+module k8s.io/cloud-provider-openstack
+
+go 1.13
+
+require (
+	github.com/NYTimes/gziphandler v1.1.1 // indirect
```
Comment (ramineni) [inline go.mod]: Is this file automatically generated or any manual changes of versions done?

File: Makefile [hunk 3]
```diff
@@ -35,10 +34,6 @@ TAGS      :=
 LDFLAGS   := "-w -s -X 'k8s.io/cloud-provider-openstack/pkg/version.Version=${VERSION}'"
 REGISTRY ?= k8scloudprovider
 
-ifneq ("$(DEST)", "$(PWD)")
```
Comment (ramineni) [inline Makefile]: we still need to run make from the base dir of project , why is this removed?

Comment (kayrus) [reply Makefile]: I doesn't matter anymore. `go` automatically detects the `go.mod` inside the current directory and builds packages independently on the path.

File: go.mod (hunk 2 above)
Comment (ramineni) [reply go.mod]: Ok, could you clarify one thing (sorry , still getting hang on using gomod), how the master branch things taken care, like we need to update from time to time , running go mod tidy or something ? 
And one more thing, doesnt vendor directory need to created again? because previous vendor was based on dep

File: Makefile (hunk 3 above)
Comment (ramineni) [reply Makefile]: But for make its required , the below error is for make right, for make to run we need to be in base dir

File: go.mod (hunk 2 above)
Comment (kayrus) [reply go.mod]: With go modules go automatically downloads all the dependencies into the go system cache, there is no need to maintain the `vendor` directory. `go mod vendor` makes sense only for projects, which commit the `vendor` dir inside the repo.

File: Makefile (hunk 3 above)
Comment (kayrus) [reply Makefile]: `make` command is usually executed within the current directory. This check is not needed, besides it [confuses end users](https://github.com/kubernetes/cloud-provider-openstack/issues/741#issuecomment-537359670), when they want to build the binary from source

File: .zuul.yaml (hunk 1 above)
Comment (kayrus) [reply .zuul.yaml]: Added

File: go.mod (hunk 2 above)
Comment (kayrus) [reply go.mod]: If you want to update dependencies, then you usually set the new dependency version in the `go.mod` and run `go mod download`. It is also possible to run `go mod download %dep_url%@version`, but sometimes it doesn't work. Then you run `go mod tidy`

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	k8s.io/klog v0.4.0
+	k8s.io/kube-openapi v0.0.0-20190918143330-0270cf2f1c1d // indirect
+	k8s.io/kubernetes v1.16.0
+	k8s.io/utils v0.0.0-20190801114015-581e00157fb1
+	sigs.k8s.io/sig-storage-lib-external-provisioner v0.0.0-20190807214443-c525773885fc
+)
+
+replace (
+	k8s.io/api => k8s.io/kubernetes/staging/src/k8s.io/api v0.0.0-20190913145653-2bd9643cee5b
+	k8s.io/apiextensions-apiserver => k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver v0.0.0-20190913145653-2bd9643cee5b
```
Comment (adisky) [inline go.mod]: @kayrus how the constraints and override like below taken care here, sorry i am also new to go modules
https://github.com/kubernetes/cloud-provider-openstack/blob/master/Gopkg.toml#L79

**Comment by adisky** [comment]
@kayrus please check why `pull-cloud-provider-openstack-check` is failing? rest all jobs have now fix in openlab zuul

File: Makefile [hunk 5]
```diff
@@ -47,123 +42,111 @@ $(GOBIN):
 
 work: $(GOBIN)
 
-depend: work
```
Comment (zetaab) [inline Makefile]: why depend is removed? Imo we should have way to update components like "previously" see my PR https://github.com/kubernetes/cloud-provider-openstack/pull/816 about depend things

**Comment by zetaab** [comment]
also this is quite old PR already. Could you please rebase against master that we can see that everything will build

**Comment by kayrus** [comment]
> also this is quite old PR already. Could you please rebase against master that we can see that everything will build

@zetaab hence it is old, it is rebased when needed. Last time I rebased it yesterday.

File: Makefile (hunk 5 above)
Comment (adisky) [reply Makefile]: @kayrus removing depend causing changes at many place, @zetaab in your PR too depend-update is not doing the same thing, for go modules i don't think there is a requirement for separate update command

Comment (adisky) [reply Makefile]: @kayrus with your PR we not creating any vendor directory before build? i see there is no requirement but @zetaab you are doing this any particular reason for that?

File: hack/verify-import-boss.sh [hunk 6]
```diff
@@ -22,4 +22,7 @@ if ! which import-boss > /dev/null; then
   exit 1
 fi
 
-import-boss -i k8s.io/cloud-provider-openstack/... --verify-only
+# prepare the vendor dir
+go mod vendor
```
Comment (adisky) [inline hack/verify-import-boss.sh]: why using  `go mod vendor` here??

Comment (kayrus) [reply hack/verify-import-boss.sh]: because it required by `import-boss` and only during tests. I don't know what is the purpose of using the `import-boss`

File: Makefile (hunk 5 above)
Comment (kayrus) [reply Makefile]: @adisky there is no requirement in using the vendor dir, except the `import-boss`  for tests. more details are here: https://github.com/golang/go/wiki/Modules#can-i-control-when-gomod-gets-updated-and-when-the-go-tools-use-the-network-to-satisfy-dependencies

Comment (zetaab) [reply Makefile]: I did vendor folder because that is how I have always done things. However, it is same to me is there vendor folder or not. The main thing is to get this PR forward. :)

Comment (adisky) [reply Makefile]: @zetaab I am ok with the changes, let the test pass and also fine in both ways having depend or not.

**Comment by adisky** [comment]
@kayrus Thanks!!
/lgtm
/assign @lingxiankong @ramineni @zetaab

**Comment by kayrus** [comment]
Any other concerns? I tested this in docker container as well, and it works perfectly:

```
docker run -ti --rm -v "${PWD}":/go/src/k8s.io/cloud-provider-openstack -w /go/src/k8s.io/cloud-provider-openstack golang:latest make build
```

**Comment by kayrus** [comment]
@lingxiankong @ramineni @zetaab @adisky any news on this PR?

**Comment by zetaab** [comment]
from my point of view it is enough if everything works :)
/approve

someone else can add lgtm
